src/v/kafka/server/group_metadata_migration.cc

```diff
@@ -0,0 +1,691 @@
+
+    // create consumer offsets topic based on the old configuration
+    while (!topics.contains(
+      model::kafka_consumer_offsets_nt, model::partition_id{0})) {
+        if (_controller.is_raft0_leader()) {
+            vlog(mlog.info, "creating consumer offsets topic");
+            co_await create_consumer_offsets_topic(
+              _controller,
+              std::move(*group_topic_assignment),
+              default_deadline());
+            continue;
+        }
+
+        co_await ss::sleep_abortable(
+          default_timeout, _controller.get_abort_source().local());
+    }
+
+    vlog(mlog.info, "waiting for preparing barrier");
+    co_await feature_manager().barrier(preparing_barrier_tag);
+
+    absl::node_hash_map<ss::shard_id, std::vector<model::ntp>> shard_ntps;
+    for (int16_t p_id = 0; p_id < static_cast<int16_t>(partitions); ++p_id) {
+        model::ntp ntp(
+          model::kafka_group_nt.ns,
+          model::kafka_group_nt.tp,
+          model::partition_id(p_id));
+
+        auto shard_id = _controller.get_shard_table().local().shard_for(ntp);
+
+        if (shard_id) {
+            shard_ntps[*shard_id].push_back(std::move(ntp));
+        }
+    }
+
+    for (auto& [shard, ntps] : shard_ntps) {
+        ssx::spawn_with_gate(
+          _partitions_gate, [this, shard = shard, ntps = ntps]() mutable {
+              vlog(
+                mlog.info,
+                "dispatching {} ntps migration on shard: {}",
+                ntps.size(),
+                shard);
+              return ss::smp::submit_to(
+                shard,
+                [&pm = _controller.get_partition_manager(),
+                 &st = _controller.get_shard_table(),
+                 &as = _controller.get_abort_source(),
+                 ntps = std::move(ntps)]() mutable -> ss::future<> {
+                    return dispatch_ntps_migration(std::move(ntps), pm, st, as);
+                });
+          });
+    }
+
+    co_await _partitions_gate.close();
+
+    vlog(mlog.info, "finished migrating kafka_internal/group partitions");
+}
+
+ss::future<> group_metadata_migration::start(ss::abort_source& as) {
+    // if version is overriden for test purposes - skip migration
+    if (
+      feature_table().get_latest_logical_version()
+      < cluster::cluster_version{2}) {
+        co_return;
+    }
+    if (feature_table().is_active(cluster::feature::consumer_offsets)) {
+        feature_manager().exit_barrier(preparing_barrier_tag);
+        feature_manager().exit_barrier(active_barrier_tag);
+        co_return;
+    }
+
+    // if no group topic is present just make feature active
+    if (!_controller.get_topics_state().local().contains(
+          model::kafka_group_nt, model::partition_id{0})) {
+        vlog(
+          mlog.info,
+          "kafka_internal/group topic does not exists, activating "
+          "consumer_offsets feature");
+        try {
+            co_await activate_feature(as);
```

Comment:
Should this be specific to the controller leader?  It seems like at the moment it'll get called  on all nodes and generate some spurious log messages about not_leader errors on other nodes.

Reply:
i will add a check to the while loop, it has to be executed on all of the nodes until the feature become active as it would fail in case of leadership change